### PR TITLE
Adding publicRooms endpoint to federation API (#626)

### PR DESCRIPTION
**Pull request by anstue**
Hi, I added the /publicRooms Endpoint to the federation server. #626

Let me know what you think.

I also tried to run SyTest with the docker image matrixdotorg/sytest-dendrite:latest, but after a few hundred tests I got the following error: 

```
+ TEST_STATUS=1
+ /src/show-expected-fail-tests.sh results.tap /src/testfile
+ mkdir -p /logs
+ cp results.tap /logs/results.tap
+ rsync --ignore-missing-args --min-size=1B -av server-0 server-1 /logs --include '*/' '--include=*.log.*' '--include=*.log' '--exclude=*'
sending incremental file list
server-0/
server-0/dendrite-logs/
server-0/dendrite-logs/Monolith.log
server-1/

sent 38,462,469 bytes  received 47 bytes  76,925,032.00 bytes/sec
total size is 38,452,881  speedup is 1.00
+ mkdir -p /logs/sytest
+ perl ./tap-to-junit-xml.pl --puretap --input=/logs/results.tap --output=/logs/sytest/results.xml SyTest
+ exit 1
```

### CODE REVIEW

**Comment by anoadragon453** [comment]
Thanks for your PR! I'll review it in a moment.

As for your docker output, could you post the logs a little earlier? These logs just tell me something failed during the `run-tests.pl` command, but not what.

File: go.mod [hunk 1]
```diff
@@ -2,66 +2,46 @@ module github.com/matrix-org/dendrite
 
 require (
```
Comment (anoadragon453) [inline go.mod]: Could you remove `go.mod` and `go.sum` from this PR please? This PR shouldn't require an update to the dependencies.

**Comment by anstue** [comment]
Sure. I'll forgot about linting

Docker output:
```
tests/90jira/SYN-606.pl:
    Test 764 Guest user can call /events on another world_readable room (SYN-606)...
    Test 765 Real user can call /events on another world_readable room (SYN-606)...
tests/90jira/SYN-627.pl:
    Test 766 Events come down the correct room...
[server]: time="2019-10-04T18:52:05.681721881Z" level=info msg="received event from roomserver" func="onNewRoomEvent\n\t" file=" [/home/uni/git/dendrite/syncapi/consumers/roomserver.go:105]" event_id="$8CIATCDUO52BydF8:localhost:8800" room_id="!rj7NVR3GabqPuTLA:localhost:8800"
[server]: time="2019-10-04T18:52:05.696818908Z" level=info msg="Responding (47 bytes)" func="respond\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:159]" code=200 req.id=zSGQtVwy2Q0S req.method=PUT req.path="/_matrix/client/r0/rooms/!rj7NVR3GabqPuTLA:localhost:8800/send/m.room.message/565"
[server]: time="2019-10-04T18:52:05.706732476Z" level=info msg="Sending event" func="SendEvent\n\t" file=" [/home/uni/git/dendrite/federationsender/queue/queue.go:62]" destinations="[]" event="$8CIATCDUO52BydF8:localhost:8800"
[server]: time="2019-10-04T18:52:05.707812866Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:111]" req.id=xJWCdiwBD1jS req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.709206222Z" level=info msg="Incoming /sync request" func="OnIncomingSyncRequest\n\t" file=" [/home/uni/git/dendrite/syncapi/sync/requestpool.go:64]" req.id=xJWCdiwBD1jS req.method=GET req.path=/_matrix/client/r0/sync since="<nil>" timeout=0s userID="@anon-20191004_183924-810:localhost:8800"
[server]: time="2019-10-04T18:52:05.858028705Z" level=info msg="Responding (54449 bytes)" func="respond\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:159]" code=200 req.id=xJWCdiwBD1jS req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.866575057Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:111]" req.id=vzbiGG2RrLwY req.method=PUT req.path="/_matrix/client/r0/rooms/!axcSM7EiV5Ourxu3:localhost:8800/send/m.room.message/566"
[server]: time="2019-10-04T18:52:05.940368774Z" level=info msg="received event from roomserver" func="onMessage\n\t" file=" [/home/uni/git/dendrite/federationsender/consumers/roomserver.go:93]" event_id="$hk9IpbgpDwQcHP7P:localhost:8800" room_id="!axcSM7EiV5Ourxu3:localhost:8800" send_as_server="localhost:8800"
[server]: time="2019-10-04T18:52:05.941787299Z" level=info msg="appservice received an event from roomserver" func="onMessage\n\t" file=" [/home/uni/git/dendrite/appservice/consumers/roomserver.go:102]" event_id="$hk9IpbgpDwQcHP7P:localhost:8800" room_id="!axcSM7EiV5Ourxu3:localhost:8800" type=m.room.message
[server]: time="2019-10-04T18:52:05.943047797Z" level=info msg="received event from roomserver" func="onMessage\n\t" file=" [/home/uni/git/dendrite/clientapi/consumers/roomserver.go:92]" event_id="$hk9IpbgpDwQcHP7P:localhost:8800" room_id="!axcSM7EiV5Ourxu3:localhost:8800" type=m.room.message
[server]: time="2019-10-04T18:52:05.945813569Z" level=info msg="received event from roomserver" func="onNewRoomEvent\n\t" file=" [/home/uni/git/dendrite/syncapi/consumers/roomserver.go:105]" event_id="$hk9IpbgpDwQcHP7P:localhost:8800" room_id="!axcSM7EiV5Ourxu3:localhost:8800"
[server]: time="2019-10-04T18:52:05.954790418Z" level=info msg="Sending event" func="SendEvent\n\t" file=" [/home/uni/git/dendrite/federationsender/queue/queue.go:62]" destinations="[]" event="$hk9IpbgpDwQcHP7P:localhost:8800"
[server]: time="2019-10-04T18:52:05.964986167Z" level=info msg="Responding (47 bytes)" func="respond\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:159]" code=200 req.id=vzbiGG2RrLwY req.method=PUT req.path="/_matrix/client/r0/rooms/!axcSM7EiV5Ourxu3:localhost:8800/send/m.room.message/566"
[server]: time="2019-10-04T18:52:05.971240017Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:111]" req.id=zl8zSGj0LTs8 req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.971843540Z" level=info msg="Incoming /sync request" func="OnIncomingSyncRequest\n\t" file=" [/home/uni/git/dendrite/syncapi/sync/requestpool.go:64]" req.id=zl8zSGj0LTs8 req.method=GET req.path=/_matrix/client/r0/sync since=2958_6 timeout=1s userID="@anon-20191004_183924-810:localhost:8800"
[server]: time="2019-10-04T18:52:05.979540281Z" level=warning msg="Didn't call Close on UserStreamListener" func="func1\n\t" file=" [/home/uni/git/dendrite/syncapi/sync/userstream.go:79]" req.id=zl8zSGj0LTs8 req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.985476107Z" level=info msg="Responding (559 bytes)" func="respond\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:159]" code=200 req.id=zl8zSGj0LTs8 req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.991363392Z" level=info msg="Incoming request" func="RequestWithLogging\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:111]" req.id=pW0jSvtU9z4O req.method=GET req.path=/_matrix/client/r0/sync
[server]: time="2019-10-04T18:52:05.992122416Z" level=info msg="Incoming /sync request" func="OnIncomingSyncRequest\n\t" file=" [/home/uni/git/dendrite/syncapi/sync/requestpool.go:64]" req.id=pW0jSvtU9z4O req.method=GET req.path=/_matrix/client/r0/sync since=2929_6 timeout=0s userID="@anon-20191004_183924-810:localhost:8800"
[server]: time="2019-10-04T18:52:06.007916657Z" level=info msg="Responding (13319 bytes)" func="respond\n\t" file=" [/home/uni/go/pkg/mod/github.com/matrix-org/util@v0.0.0-20171127121716-2e2df66af2f5/json.go:159]" code=200 req.id=pW0jSvtU9z4O req.method=GET req.path=/_matrix/client/r0/sync
+ TEST_STATUS=1
+ /src/show-expected-fail-tests.sh results.tap /src/testfile
```

**Comment by anstue** [comment]
I got the results.tap file from the logs:
```
...
not ok 764 (expected fail) Guest user can call /events on another world_readable room (SYN-606) # TODO expected fail
# Started: 2019-10-05 07:10:52.693
# Ended: 2019-10-05 07:10:52.703
# fixture failed - HTTP Request failed ( 401 Unauthorized https://localhost:8800/_matrix/client/r0/register?kind=guest ) from POST https://localhost:8800/_matrix/client/r0/register?...
# {"flows":[{"stages":["m.login.dummy"]}],"completed":[],"params":{},"session":"xWcbJDvWkhtTgRNn3LnQ4uyF"}
not ok 765 (expected fail) Real user can call /events on another world_readable room (SYN-606) # TODO expected fail
# Started: 2019-10-05 07:10:52.703
# Ended: 2019-10-05 07:10:54.497
# HTTP Request failed ( 403 Forbidden https://localhost:8800/_matrix/client/r0/rooms/!BeyOigI6u8ZfrlCC:localhost:8800/initialSync?access_token=VY3rqP2MjcDXW0IhBchsp3EjWwpkQrjIboLN9wSmhi0 ) from GET https://localhost:8800/_matrix/client/r0/rooms/!BeyOigI6u8ZfrlCC:localhost:8800/initialSync?...
# {"errcode":"M_GUEST_ACCESS_FORBIDDEN","error":"Guest access not implemented"}
# 0.144347: Registered new user @anon-20191005_065840-807:localhost:8800
# 0.145939: Registered new user @anon-20191005_065840-805:localhost:8800
# 0.152851: Registered new user @anon-20191005_065840-808:localhost:8800
# 0.227432: Registered new user @anon-20191005_065840-809:localhost:8800
# 1.416482: room_id=!jXdcHLFiJc32hIiH:localhost:8800
# 1.520102: room_id=!BeyOigI6u8ZfrlCC:localhost:8800
ok 766 Events come down the correct room
# Killing process 209
#** Homeserver process failed - code=2
1..766
```

**Comment by anoadragon453** [comment]
Tests that have `# TODO expected fail` next to it are expected to fail, as they are not marked as passing in Dendrite's [Sytest whitelist file](https://github.com/matrix-org/dendrite/blob/master/testfile), so no need to worry about those.

Dendrite master is currently failing on the following tests however:

* Alias creators can delete alias with no ops
* Can add tag
* Can remove tag

We're working to fix those shortly.

**Comment by anoadragon453** [comment]
CI has now been fixed by https://github.com/matrix-org/dendrite/pull/804

Can you merge this PR forward to latest master so it will get these fixes as well please?

**Comment by anstue** [comment]
Sure

**Comment by anstue** [comment]
Anything left or can it be merged?

**Comment by Kegsay** [comment]
Thanks for your work on this. This issue was resolved recently when https://github.com/matrix-org/dendrite/pull/923 got merged.

**Pull request closed without merging**